### PR TITLE
Stage2 trs stake

### PR DESCRIPTION
**Pull request by askarovit**

### CODE REVIEW

File: core/service/transaction/stake.ts [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        buff.writeInt32LE(trs.asset.stakeOrder.startTime, offset);
+        offset += BUFFER.LENGTH.UINT32;
+        buff.writeInt8(trs.asset.airdropReward.withAirdropReward ? 1 : 0, offset);
+        offset += BUFFER.LENGTH.BYTE;
+        BUFFER.writeUInt64LE(buff, trs.asset.airdropReward.totalReward || 0, offset);
+
+        const referralBuffer = Buffer.alloc(BUFFER.LENGTH.INT64 + BUFFER.LENGTH.INT64);
+        offset = 0;
+        if (trs.asset.airdropReward.sponsors && Object.keys(trs.asset.airdropReward.sponsors).length > 0) {
+            const address = Object.keys(trs.asset.airdropReward.sponsors)[0];
```
Comment (axelsheva) [inline core/service/transaction/stake.ts]: sponsors are `Array<string>`?

Comment (axelsheva) [inline core/service/transaction/stake.ts]: the address must be Address type

File: core/service/transaction/stake.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        if (trs.asset.stakeOrder.stakedAmount <= 0 && config.constants.STAKE_VALIDATE.AMOUNT_ENABLED) {
+            errors.push('Invalid transaction amount');
+        }
+
+        if ((trs.asset.stakeOrder.stakedAmount % 1) !== 0 && config.constants.STAKE_VALIDATE.AMOUNT_ENABLED) {
+            errors.push('Invalid stake amount: Decimal value');
+        }
+
+        const airdropCheck: ResponseEntity<any> = await verifyAirdrop(trs, trs.asset.stakeOrder.stakedAmount, sender);
+        if (airdropCheck.errors && airdropCheck.errors.length > 0 && config.constants.STAKE_VALIDATE.AIRDROP_ENABLED) {
```
Comment (axelsheva) [inline core/service/transaction/stake.ts]: Should be `if (!airdropCheck.success)`

File: core/service/transaction/stake.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        return AccountRepo.updateBalanceByAddress(trs.senderAddress, totalAmount * (-1));
+    }
+
+    async undoUnconfirmed(trs: Transaction<IAssetStake>, sender: Account): Promise<ResponseEntity<void>> {
+        const fee: number = this.calculateFee(trs);
+        const totalAmount: number = fee + trs.asset.stakeOrder.stakedAmount;
+        return AccountRepo.updateBalanceByAddress(trs.senderAddress, totalAmount);
+    }
+
+    async apply(trs: Transaction<IAssetStake>):  Promise<ResponseEntity<void>> {
```
Comment (axelsheva) [inline core/service/transaction/stake.ts]: double space

File: core/service/transaction/stake.ts [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        await sendAirdropReward(trs);
```
Comment (axelsheva) [inline core/service/transaction/stake.ts]: I think it should be in `applyUnconfirmed`. `apply` should be empty now

File: core/util/reward.ts [hunk 5]
```diff
@@ -151,7 +151,7 @@ async function applyUnstake(orders: Array<Stake>, voteTransaction: Transaction<I
     AccountRepo.updateBalanceByAddress(voteTransaction.senderAddress, voteTransaction.asset.unstake);
 }
 
-async function sendAirdropReward(trs: Transaction<IAssetVote>): Promise<void> {
+export async function sendAirdropReward(trs: Transaction<IAssetVote | IAssetStake>): Promise<void> {
```
Comment (axelsheva) [inline core/util/reward.ts]: Stake trs has airdrop reward?

File: core/service/transaction/stake.ts [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+import {
+    getAirdropReward,
+    verifyAirdrop,
+} from 'core/util/reward';
+
+import {ITableObject} from 'core/util/common';
+
+class TransactionStakeService implements ITransactionService<IAssetStake> {
+
+    async create(trs: Transaction<IAssetStake>, data?: IAssetStake ): Promise<IAssetStake> {
```
Comment (Verbalist) [inline core/service/transaction/stake.ts]: sync

File: core/util/reward.ts [hunk 7]
```diff
@@ -151,7 +151,7 @@ function applyUnstake(orders: Array<Stake>, voteTransaction: Transaction<IAssetV
     AccountRepo.updateBalanceByAddress(voteTransaction.senderAddress, voteTransaction.asset.unstake);
 }
 
-function sendAirdropReward(trs: Transaction<IAssetVote>): void {
+async function sendAirdropReward(trs: Transaction<IAssetVote>): Promise<void> {
```
Comment (Verbalist) [inline core/util/reward.ts]: sync

**Pull request merged**